cli/monitor.py

```diff
@@ -0,0 +1,700 @@
+            event (saltevent.SaltEvent): the step object
+        """
+        assert self._executing
+
+        if self.current_step >= len(self._steps):
+            return None
+
+        if isinstance(event, RetRunnerEvent):
+            curr_step = self._steps[self.current_step]
+            if curr_step.jid and curr_step.jid == event.jid:
+                curr_step.finish(event)
+                self.current_step += 1
+                return self._steps[self.current_step-1]
+
+        elif isinstance(event, RetJobEvent):
+            curr_step = self._steps[self.current_step]
+            if curr_step.jid and curr_step.jid == event.jid:
+                curr_step.finish(event)
+                if curr_step.finished:
+                    self.current_step += 1
+                return curr_step
+
+        else:
+            assert False
+
+        if not self._enable_dynamic:
+            return None
+
+        # this step is not part of stage parsed steps
+        if isinstance(event, RetRunnerEvent):
+            if event.jid in self._dynamic_steps:
+                step = self._dynamic_steps[event.jid]
+                step.finish(event)
+                return step
+        elif isinstance(event, RetJobEvent):
+            if event.jid in self._dynamic_steps:
+                step = self._dynamic_steps[event.jid]
+                step.finish(event)
+                return step
+
+        return None
+
+    def state_result_step(self, event):
+        """
+        Consumes teh current step
+        Args:
+            event (saltevent.StateResultEvent): the state result event
+        """
+        assert self._executing
+
+        if self.current_step >= len(self._steps):
+            return None
+
+        curr_step = self._steps[self.current_step]
+        assert not curr_step.finished
+
+        if event.jid == curr_step.jid:
+            curr_step.state_result(event)
+            return curr_step
+
+        return None
+
+    def check_if_current_step_will_run(self):
+        assert self._executing
+
+        if self.current_step >= len(self._steps):
+            return None
+
+        curr_step = self._steps[self.current_step]
+        for dep in curr_step.step.on_success_deps:
+            for i in range(1, self.current_step+1):
+                dep_step = self._steps[self.current_step-i]
+                if dep_step.step.desc == dep.desc:
+                    if not dep_step.success:
+                        curr_step.skipped = True
+                        self.current_step += 1
+                        return curr_step
+        for dep in curr_step.step.on_fail_deps:
+            for i in range(1, self.current_step+1):
+                dep_step = self._steps[self.current_step-i]
+                if dep_step.step.desc == dep.desc:
+                    if dep_step.success:
+                        curr_step.skipped = True
+                        self.current_step += 1
+                        return curr_step
+
+        return None
```

Comment:
Why are we returning None, can't we return True/False instead?

Reply:
Because we are returning an object in the above returns, and I think we should not mix different types for a return function.